docs/madr/decisions/031-automatic-rechable-services.md

```diff
@@ -0,0 +1,263 @@
+Based on that information we could build a reverse graph containing information who can talk to whom and based on that we could trim the configs created (which would be equivalent to populating reachable services field).
+
+So MTPs like this:
+
+```yaml
+apiVersion: kuma.io/v1alpha1
+kind: MeshTrafficPermission
+metadata:
+  namespace: kuma-system
+  name: mtp-b
+spec:
+  targetRef:
+    kind: MeshService
+    name: d
+  from:
+    - targetRef:
+        kind: MeshService
+        name: a
+      default:
+        action: Allow
+    - targetRef:
+        kind: MeshService
+        name: b
+      default:
+        action: Allow
+    - targetRef:
+        kind: MeshService
+        name: c
+      default:
+        action: Allow
+```
+
+Would translate into:
+
+```mermaid
+graph TD
+    A-->D
+    B-->D
+    C-->D
+```
+
+Let's consider other `targetRef` with allow types in combination with `top`/`from` level:
+1. `Mesh/Mesh` - this won't help us because that's the default, anyone can talk to anyone and we can't
+
+2. `MeshService/Mesh`  
+
+```yaml
+apiVersion: kuma.io/v1alpha1
+kind: MeshTrafficPermission
+metadata:
+  namespace: kuma-system
+  name: mtp-ms-to-mesh
+spec:
+  targetRef:
+    kind: MeshService
+    name: a
+  from:
+    - targetRef:
+        kind: Mesh
+      default:
+        action: Allow
+```
+
+```mermaid
+graph TD
+    B-->A
+    C-->A
+    D-->A
+```
+
+We could either define an edge between every service to that service or have a special list that contains allIncomingAllowed.
+Services from that list would always be added as an outgoing edge.
+
+- `Mesh/MeshService`
+
+```yaml
+apiVersion: kuma.io/v1alpha1
+kind: MeshTrafficPermission
+metadata:
+  namespace: kuma-system
+  name: mtp-mesh-to-ms
+spec:
+  targetRef:
+    kind: Mesh
+  from:
+    - targetRef:
+        kind: MeshService
+        name: a
+      default:
+        action: Allow
+```
+
+```mermaid
+graph TD
+    A-->B
+    A-->C
+    A-->D
+```
+
+This is simply `A` can talk to anyone, this could be either adding and outgoing edge from this service to every other service or having it on a list called allOutgoing and not bothering to trim the configuration.
+
+3. `MeshService/MeshService` - already talked about, this is the easy case
+
+4. `MeshServiceSubset` 
+
+Unfortunately there is one option here,
+we need to treat `MeshServiceSubset` just like `MeshService`.
+This is because we can't trim individual endpoints because the data is not available at that point, and
+we have to accept the fact that we will miss out on (small) parts of the config that could be trimmed.
+
+5. `MeshSubset`
+
+This one is a bit more tricky because the potential radious is a bit bigger.
+The approach has to be the same as in `MeshServiceSubset`, meaning: treat `MeshSubset` as just `Mesh`.
+
+There is a case that makes sense using `k8s.kuma.io/namespace: xyz` to allow only traffic from a specific namespace,
+so this needs to be documented that it won't improve the performance.
```

Comment:
I feel that we take this a little upside down. What we do is when we generate the configuration of a DPP we want to figure which clusters we have access to. At this time we should be able to look at all the MTP one by one and the list of services.
So we can't manage a `MeshSubset` at the top level but we can at the `from` level right?

So in the end I think the algorithm is:

- take all MTPs
- filter out all MTP with topLevel targetRef of MeshSubset
- for each service in the mesh keep only those who have an MTP which have access to ANY endpoint in this service from the current DPP

no?

Reply:
> So we can't manage a MeshSubset at the top level but we can at the from level right?

I think you're right, but I guess @jakubdyszkiewicz would be the best to confirm this

Reply:
see https://github.com/kumahq/kuma/pull/8084#discussion_r1371331838

Reply:
I updated the whole section